returnn/tf/layers/rec.py

```diff
@@ -2904,6 +2904,12 @@ def output_can_move_out(layer, _collocated=None):
         if layer in other_layer.dependencies:
           return False
         if other_layer.name in layer.collocate_with:
+          # layer is collocated with other_layer.
+          # we can move out layer if other_layer has no dependencies on any layer in loop but on the sibling layers
+          # that themselves are collocated with other_layer
+          other_layer_collocated_with = [
+            sibling for sibling in layers_in_loop if other_layer.name in sibling.collocate_with]
+          _collocated.update(other_layer_collocated_with)
```

Comment:
I think the logic direction is wrong. `x in y.collocate_with` says that `y` should be collocated with `x`. So if `x` must be inside, `y` must be as well. But it has no implications on `x`. `x` could be inside or outside.

Reply:
The variable name is misleading. I wanted this set to be the set of layers that should be collocated with `other_layer` (with the notion that they are "siblings" of `layer`).
`layers_collocated_with_other_layer` or so would be better, I'll change it.

Reply:
> I think the logic direction is wrong. `x in y.collocate_with` says that `y` should be collocated with `x`. So if `x` must be inside, `y` must be as well. But it has no implications on `x`. `x` could be inside or outside.

Sorry, I was partly wrong here. It has implications on `x` as well. It goes both ways. It means they belong together. And if `collocate_with` is incorrectly set, i.e. not possible to have both collocated together, this should be an error.

So I think the logic of your code is partly correct. However, it should also check `can_move_out` for all the other collocated layers. So it is wrong to just directly call `_collocated.update`.